pkg/aggregator/aggregator.go

```diff
@@ -51,71 +71,89 @@ func FilterSpecByPaths(sp *spec.Swagger, keepPathPrefixes []string) {
-	orgPaths := sp.Paths
-	if orgPaths == nil {
-		return
-	}
-	sp.Paths = &spec.Paths{
-		VendorExtensible: orgPaths.VendorExtensible,
+	ret := *sp
+	ret.Paths = &spec.Paths{
+		VendorExtensible: sp.Paths.VendorExtensible,
 		Paths:            map[string]spec.PathItem{},
 	}
-	for path, pathItem := range orgPaths.Paths {
+	for path, pathItem := range sp.Paths.Paths {
 		if !prefixes.HasPrefix(path) {
 			continue
 		}
-		sp.Paths.Paths[path] = pathItem
+		ret.Paths.Paths[path] = pathItem
 	}
 
 	// Walk all references to find all definition references.
-	usedDefinitions := usedDefinitionForSpec(sp)
+	usedDefinitions := usedDefinitionForSpec(&ret)
 
 	// Remove unused definitions
-	orgDefinitions := sp.Definitions
-	sp.Definitions = spec.Definitions{}
-	for k, v := range orgDefinitions {
+	ret.Definitions = spec.Definitions{}
+	for k, v := range sp.Definitions {
 		if usedDefinitions[k] || !initialUsedDefinitions[k] {
-			sp.Definitions[k] = v
+			ret.Definitions[k] = v
 		}
 	}
+
+	return &ret
 }
 
-func renameDefinition(s *spec.Swagger, old, new string) {
+// renameDefinition renames references, without mutating the input.
+// The output might share data structures with the input.
+func renameDefinition(s *spec.Swagger, old, new string) *spec.Swagger {
 	oldRef := definitionPrefix + old
 	newRef := definitionPrefix + new
-	walkOnAllReferences(func(ref spec.Ref) spec.Ref {
+
+	if _, ok := s.Definitions[old]; !ok {
+		return s
+	}
+
+	ret := &spec.Swagger{}
+	*ret = *s
+
+	ret = replaceReferences(func(ref *spec.Ref) *spec.Ref {
 		if ref.String() == oldRef {
-			return spec.MustCreateRef(newRef)
+			ret := spec.MustCreateRef(newRef)
+			return &ret
 		}
 		return ref
-	}, s)
-	// Make sure we don't assign to nil map
-	if s.Definitions == nil {
-		s.Definitions = spec.Definitions{}
+	}, ret)
+
+	renamedDefinitions := make(spec.Definitions, len(ret.Definitions))
+	for k, v := range ret.Definitions {
+		if k == old {
+			k = new
+		}
+		renamedDefinitions[k] = v
 	}
-	s.Definitions[new] = s.Definitions[old]
-	delete(s.Definitions, old)
+	ret.Definitions = renamedDefinitions
+
+	return ret
 }
 
-// MergeSpecsIgnorePathConflict is the same as MergeSpecs except it will ignore any path
+// MergePathsAndDefinitionsIgnorePathConflict is the same as MergePathAndDefinitions except it will ignore any path
 // conflicts by keeping the paths of destination. It will rename definition conflicts.
-func MergeSpecsIgnorePathConflict(dest, source *spec.Swagger) error {
+// The source is not mutated.
+func MergePathsAndDefinitionsIgnorePathConflict(dest, source *spec.Swagger) error {
 	return mergeSpecs(dest, source, true, true)
 }
 
-// MergeSpecsFailOnDefinitionConflict is differ from MergeSpecs as it fails if there is
+// MergePathsAndDefinitionsFailOnDefinitionConflict is differ from MergePathAndDefinitions as it fails if there is
 // a definition conflict.
-func MergeSpecsFailOnDefinitionConflict(dest, source *spec.Swagger) error {
```

Comment:
renaming a public method is backward incompatible.

Reply:
fixed